File: src/main/java/mediathek/tool/cellrenderer/CellRendererDownloads.java
package mediathek.tool.cellrenderer;

import com.formdev.flatlaf.extras.FlatSVGIcon;
import mediathek.config.MVColor;
import mediathek.controller.starter.Start;
import mediathek.daten.DatenDownload;
import mediathek.swing.IconUtils;
import mediathek.tool.SVGIconUtilities;
import mediathek.tool.table.MVTable;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.kordamp.ikonli.fontawesome6.FontAwesomeRegular;
import org.kordamp.ikonli.fontawesome6.FontAwesomeSolid;
import org.kordamp.ikonli.swing.FontIcon;

import javax.swing.*;
import javax.swing.border.Border;
import java.awt.*;

public class CellRendererDownloads extends CellRendererBaseWithStart {
    private final static String DOWNLOAD_STARTEN = "Download starten";
    private final static String DOWNLOAD_LOESCHEN = "Download aus Liste entfernen";
    private final static String DOWNLOAD_STOPPEN = "Download stoppen";
    private final static String DOWNLOAD_ENTFERNEN = "Download entfernen";
    private final static String PLAY_DOWNLOADED_FILM = "gespeicherten Film abspielen";
    private static final Logger logger = LogManager.getLogger(CellRendererDownloads.class);
    private final FontIcon film_start_tab;
    private final FontIcon film_start_sw_tab;
    private final Border emptyBorder = BorderFactory.createEmptyBorder(3,2,3,2);
    private final Border largeBorder = BorderFactory.createEmptyBorder(9, 2, 9, 2);
    private final JPanel panel;
    private final FontIcon download_stop_tab;
    private final FontIcon download_stop_sw_tab;
    private final FlatSVGIcon download_start_tab;
    private final Icon download_start_sw_tab;
    private final FontIcon download_clear_tab_selected;
    private final FontIcon download_clear_sw_tab;
    private final FontIcon download_del_tab_selected;
    private final FontIcon download_del_sw_tab;
    private final JProgressBar progressBar = new JProgressBar(0, 1000);

    public CellRendererDownloads() {
        download_stop_tab = FontIcon.of(FontAwesomeSolid.STOP, IconUtils.DEFAULT_SIZE, Color.WHITE);
        download_stop_sw_tab = IconUtils.of(FontAwesomeSolid.STOP);

        download_start_tab = SVGIconUtilities.createSVGIcon("icons/fontawesome/caret-down.svg");
        download_start_tab.setColorFilter(new FlatSVGIcon.ColorFilter(_ -> Color.WHITE));
        download_start_sw_tab = SVGIconUtilities.createSVGIcon("icons/fontawesome/caret-down.svg");
        download_clear_tab_selected = FontIcon.of(FontAwesomeSolid.ERASER, IconUtils.DEFAULT_SIZE, Color.WHITE);
        download_clear_sw_tab = IconUtils.of(FontAwesomeSolid.ERASER);

        download_del_tab_selected = FontIcon.of(FontAwesomeRegular.TRASH_ALT, IconUtils.DEFAULT_SIZE, Color.WHITE);
        download_del_sw_tab = IconUtils.of(FontAwesomeRegular.TRASH_ALT);

        film_start_tab = FontIcon.of(FontAwesomeSolid.PLAY, IconUtils.DEFAULT_SIZE, Color.WHITE);
        film_start_sw_tab = IconUtils.of(FontAwesomeSolid.PLAY);

        panel = new JPanel(new BorderLayout());
        panel.add(progressBar);
    }

    private void applyHorizontalAlignment(int colIndex) {
        switch (colIndex) {
            case DatenDownload.DOWNLOAD_PROGRESS, DatenDownload.DOWNLOAD_FILM_NR, DatenDownload.DOWNLOAD_NR, DatenDownload.DOWNLOAD_DATUM, DatenDownload.DOWNLOAD_ZEIT,
                    DatenDownload.DOWNLOAD_DAUER, DatenDownload.DOWNLOAD_BANDBREITE,
                    DatenDownload.DOWNLOAD_RESTZEIT -> setHorizontalAlignment(SwingConstants.CENTER);
            case DatenDownload.DOWNLOAD_GROESSE -> setHorizontalAlignment(SwingConstants.RIGHT);
        }
    }
    @Override
    public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus,
                                                   int row, int column) {
        try {
            resetComponent();
            super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column);

            final int rowModelIndex = table.convertRowIndexToModel(row);
            final int columnModelIndex = table.convertColumnIndexToModel(column);
            DatenDownload datenDownload = (DatenDownload) table.getModel().getValueAt(rowModelIndex, DatenDownload.DOWNLOAD_REF);
            final MVTable mvTable = (MVTable) table;

            if (mvTable.isLineBreak()) {
                setHorizontalAlignment(SwingConstants.LEFT);
                setVerticalAlignment(SwingConstants.TOP);

                switch (columnModelIndex) {
                    case DatenDownload.DOWNLOAD_TITEL, DatenDownload.DOWNLOAD_THEMA, DatenDownload.DOWNLOAD_URL, DatenDownload.DOWNLOAD_PROGRAMM_AUFRUF, DatenDownload.DOWNLOAD_PROGRAMM_AUFRUF_ARRAY, DatenDownload.DOWNLOAD_FILM_URL, DatenDownload.DOWNLOAD_URL_SUBTITLE, DatenDownload.DOWNLOAD_ZIEL_DATEINAME, DatenDownload.DOWNLOAD_ZIEL_PFAD, DatenDownload.DOWNLOAD_ZIEL_PFAD_DATEINAME, DatenDownload.DOWNLOAD_ABO -> {
                        var textArea = new JTextArea();
                        textArea.setLineWrap(true);
                        textArea.setWrapStyleWord(true);
                        textArea.setText(value.toString());
                        textArea.setForeground(getForeground());
                        textArea.setBackground(getBackground());
                        if (columnModelIndex == DatenDownload.DOWNLOAD_ABO) {
                            handleAboColumn(textArea, datenDownload);
                        }
                        setBackgroundColor(textArea, datenDownload.start, isSelected);
                        return textArea;
                    }
                }
            }
            else
                applyHorizontalAlignment(columnModelIndex);

            switch (columnModelIndex) {
                case DatenDownload.DOWNLOAD_PROGRESS -> {
                    if (mvTable.showSenderIcons() && !mvTable.useSmallSenderIcons) {
                        progressBar.setBorder(largeBorder);
                    } else {
                        progressBar.setBorder(emptyBorder);
                    }
                    if (datenDownload.start != null) {
                        if (1 < datenDownload.start.percent && datenDownload.start.percent < Start.PROGRESS_FERTIG) {

                            setBackgroundColor(panel, datenDownload.start, isSelected);
                            setBackgroundColor(progressBar, datenDownload.start, isSelected);

                            progressBar.setValue(datenDownload.start.percent);

                            final double progressValue = datenDownload.start.percent / 10.0;
                            progressBar.setString(Double.toString(progressValue) + '%');

                            return panel;
                        } else {
                            setText(Start.getTextProgress(datenDownload.isDownloadManager(), datenDownload.start));
                        }
                    } else {
                        setText("");
                    }
                }
                case DatenDownload.DOWNLOAD_FILM_NR -> {
                    if ((int) table.getModel().getValueAt(rowModelIndex, DatenDownload.DOWNLOAD_FILM_NR) == 0) {
                        setText("");
                    }
                }
                case DatenDownload.DOWNLOAD_ART -> {
                    switch (datenDownload.art) {
                        case DatenDownload.ART_DOWNLOAD -> setText(DatenDownload.ART_DOWNLOAD_TXT);
                        case DatenDownload.ART_PROGRAMM -> setText(DatenDownload.ART_PROGRAMM_TXT);
                    }
                }
                case DatenDownload.DOWNLOAD_QUELLE -> {
                    switch (datenDownload.quelle) {
                        case DatenDownload.QUELLE_ALLE -> setText(DatenDownload.QUELLE_ALLE_TXT);
                        case DatenDownload.QUELLE_ABO -> setText(DatenDownload.QUELLE_ABO_TXT);
                        case DatenDownload.QUELLE_BUTTON -> setText(DatenDownload.QUELLE_BUTTON_TXT);
                        case DatenDownload.QUELLE_DOWNLOAD -> setText(DatenDownload.QUELLE_DOWNLOAD_TXT);
                    }
                }
                case DatenDownload.DOWNLOAD_BUTTON_START -> handleButtonStartColumn(datenDownload, isSelected);
                case DatenDownload.DOWNLOAD_BUTTON_DEL -> handleButtonDeleteColumn(datenDownload, isSelected);
                case DatenDownload.DOWNLOAD_ABO -> handleAboColumn(datenDownload);
                case DatenDownload.DOWNLOAD_SENDER -> {
                    if (mvTable.showSenderIcons()) {
                        Dimension targetDim = getSenderCellDimension(table, row, columnModelIndex);
                        setSenderIcon(value.toString(), targetDim);
                    }
                }
                case DatenDownload.DOWNLOAD_GEO -> drawGeolocationIcons(datenDownload.film, isSelected);
            }

            if (columnModelIndex == DatenDownload.DOWNLOAD_TITEL) {
                if (datenDownload.film != null) {
                    setIndicatorIcons(table, datenDownload.film, isSelected);
                }
            }

            setBackgroundColor(this, datenDownload.start, isSelected);
        } catch (Exception ex) {
            logger.error(ex);
        }
        return this;
    }

    private void setIconsAndToolTips(DatenDownload datenDownload, Icon filmIcon,
                                     Icon downloadStartIcon, Icon downloadStopIcon) {
        if (datenDownload.start != null && !datenDownload.isDownloadManager()) {
            switch (datenDownload.start.status) {
                case Start.STATUS_FERTIG -> {
                    setIcon(filmIcon);
                    setToolTipText(PLAY_DOWNLOADED_FILM);
                }
                case Start.STATUS_ERR -> {
                    setIcon(downloadStartIcon);
                    setToolTipText(DOWNLOAD_STARTEN);
                }
                default -> {
                    setIcon(downloadStopIcon);
                    setToolTipText(DOWNLOAD_STOPPEN);
                }
            }
        } else {
            setIcon(downloadStartIcon);
            setToolTipText(DOWNLOAD_STARTEN);
        }
    }

    private void handleButtonStartColumn(final DatenDownload datenDownload, final boolean isSelected) {
        setHorizontalAlignment(SwingConstants.CENTER);
        if (isSelected) {
            setIconsAndToolTips(datenDownload, film_start_tab, download_start_tab, download_stop_tab);
        } else {
            setIconsAndToolTips(datenDownload, film_start_sw_tab, download_start_sw_tab, download_stop_sw_tab);
        }
    }

    private void handleAboColumn(JTextArea a, final DatenDownload datenDownload) {
        if (!datenDownload.arr[DatenDownload.DOWNLOAD_ABO].isEmpty()) {
            a.setForeground(MVColor.DOWNLOAD_IST_ABO.color);
        } else {
            a.setForeground(MVColor.DOWNLOAD_IST_DIREKTER_DOWNLOAD.color);
            a.setText("Download");
        }
    }

    private void handleAboColumn(final DatenDownload datenDownload) {
        setHorizontalAlignment(SwingConstants.CENTER);
        if (!datenDownload.arr[DatenDownload.DOWNLOAD_ABO].isEmpty()) {
            setForeground(MVColor.DOWNLOAD_IST_ABO.color);
        } else {
            setForeground(MVColor.DOWNLOAD_IST_DIREKTER_DOWNLOAD.color);
            setText("Download");
        }
    }

    private void handleButtonDeleteColumn(final DatenDownload datenDownload, final boolean isSelected) {
        setHorizontalAlignment(SwingConstants.CENTER);
        if (datenDownload.start != null) {
            if (datenDownload.start.status >= Start.STATUS_FERTIG) {
                setIcons(download_clear_tab_selected, download_clear_sw_tab, DOWNLOAD_ENTFERNEN, isSelected);
            } else {
                setupDownloadLoeschen(isSelected);
            }
        } else {
            setupDownloadLoeschen(isSelected);
        }
    }

    private void setIcons(Icon tab, Icon tab_sw, String text, final boolean isSelected) {
        final Icon icon;
        if (isSelected) {
            icon = tab;
        } else {
            icon = tab_sw;
        }
        setIcon(icon);
        setToolTipText(text);
    }

    private void setupDownloadLoeschen(final boolean isSelected) {
        setIcons(download_del_tab_selected, download_del_sw_tab, DOWNLOAD_LOESCHEN, isSelected);
    }
}

File: src/main/kotlin/mediathek/tool/cellrenderer/CellRendererDownloads.kt
package mediathek.tool.cellrenderer

import com.formdev.flatlaf.extras.FlatSVGIcon
import mediathek.config.MVColor
import mediathek.controller.DownloadColumn
import mediathek.controller.starter.DownloadProgressText
import mediathek.controller.starter.DownloadRunState
import mediathek.controller.starter.StartStatus
import mediathek.daten.DatenDownload
import mediathek.swing.IconUtils
import mediathek.tool.SVGIconUtilities
import mediathek.tool.table.MVTable
import org.apache.logging.log4j.LogManager
import org.kordamp.ikonli.fontawesome6.FontAwesomeRegular
import org.kordamp.ikonli.fontawesome6.FontAwesomeSolid
import org.kordamp.ikonli.swing.FontIcon
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import javax.swing.*
import javax.swing.border.Border

class CellRendererDownloads : CellRendererBaseWithStart() {
    private val filmStartIcons = rendererIconPair(
        normal = IconUtils.of(FontAwesomeSolid.PLAY),
        selected = FontIcon.of(FontAwesomeSolid.PLAY, IconUtils.DEFAULT_SIZE, Color.WHITE),
    )
    private val emptyBorder: Border = BorderFactory.createEmptyBorder(3, 2, 3, 2)
    private val largeBorder: Border = BorderFactory.createEmptyBorder(9, 2, 9, 2)
    private val progressBar = JProgressBar(0, 1000)
    private val panel = JPanel(BorderLayout()).apply {
        add(progressBar)
    }
    private val downloadStopIcons = rendererIconPair(
        normal = IconUtils.of(FontAwesomeSolid.STOP),
        selected = FontIcon.of(FontAwesomeSolid.STOP, IconUtils.DEFAULT_SIZE, Color.WHITE),
    )
    private val downloadStartIcons = rendererIconPair(
        normal = SVGIconUtilities.createSVGIcon("icons/fontawesome/caret-down.svg"),
        selected = SVGIconUtilities.createSVGIcon("icons/fontawesome/caret-down.svg").apply {
            colorFilter = FlatSVGIcon.ColorFilter { Color.WHITE }
        },
    )
    private val downloadClearIcons = rendererIconPair(
        normal = IconUtils.of(FontAwesomeSolid.ERASER),
        selected = FontIcon.of(FontAwesomeSolid.ERASER, IconUtils.DEFAULT_SIZE, Color.WHITE),
    )
    private val downloadDeleteIcons = rendererIconPair(
        normal = IconUtils.of(FontAwesomeRegular.TRASH_ALT),
        selected = FontIcon.of(FontAwesomeRegular.TRASH_ALT, IconUtils.DEFAULT_SIZE, Color.WHITE),
    )

    private fun applyHorizontalAlignment(column: DownloadColumn) {
        when (column) {
            DownloadColumn.PROGRESS,
            DownloadColumn.FILM_NUMBER,
            DownloadColumn.NUMBER,
            DownloadColumn.DATE,
            DownloadColumn.TIME,
            DownloadColumn.DURATION,
            DownloadColumn.BANDWIDTH,
            DownloadColumn.REMAINING_TIME,
                -> horizontalAlignment = CENTER

            DownloadColumn.SIZE -> horizontalAlignment = RIGHT
            else -> Unit
        }
    }

    private fun setBackgroundColor(c: Component, s: DownloadRunState?, isSelected: Boolean) {
        if (s != null) {
            val color = when (s.status) {
                StartStatus.INITIALIZED -> if (isSelected) MVColor.DOWNLOAD_WAIT_SEL.color else MVColor.DOWNLOAD_WAIT.color
                StartStatus.RUNNING -> if (isSelected) MVColor.DOWNLOAD_RUN_SEL.color else MVColor.DOWNLOAD_RUN.color
                StartStatus.FINISHED -> if (isSelected) MVColor.DOWNLOAD_FERTIG_SEL.color else MVColor.DOWNLOAD_FERTIG.color
                StartStatus.ERROR -> if (isSelected) MVColor.DOWNLOAD_FEHLER_SEL.color else MVColor.DOWNLOAD_FEHLER.color
            }
            c.background = color
        }
    }

    override fun getTableCellRendererComponent(
        table: JTable,
        value: Any?,
        isSelected: Boolean,
        hasFocus: Boolean,
        row: Int,
        column: Int,
    ): Component {
        try {
            resetComponent()
            super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column)

            val rowModelIndex = table.convertRowIndexToModel(row)
            val downloadColumn = DownloadColumn.fromIndex(table.convertColumnIndexToModel(column))
            val datenDownload = table.model.getValueAt(rowModelIndex, DownloadColumn.REF.index) as DatenDownload
            val mvTable = table as MVTable

            if (mvTable.isLineBreak()) {
                horizontalAlignment = LEFT
                verticalAlignment = TOP

                when (downloadColumn) {
                    DownloadColumn.TITLE,
                    DownloadColumn.TOPIC,
                    DownloadColumn.URL,
                    DownloadColumn.PROGRAM_INVOCATION,
                    DownloadColumn.PROGRAM_INVOCATION_ARRAY,
                    DownloadColumn.FILM_URL,
                    DownloadColumn.SUBTITLE_URL,
                    DownloadColumn.TARGET_FILE_NAME,
                    DownloadColumn.TARGET_PATH,
                    DownloadColumn.TARGET_PATH_FILE_NAME,
                    DownloadColumn.ABO,
                        -> return createTextArea(valueText(value), datenDownload, downloadColumn, isSelected)

                    else -> Unit
                }
            } else {
                applyHorizontalAlignment(downloadColumn)
            }

            when (downloadColumn) {
                DownloadColumn.PROGRESS -> renderProgressColumn(datenDownload, mvTable, isSelected)?.let { return it }
                DownloadColumn.FILM_NUMBER -> hideZeroFilmNumber(table, rowModelIndex)
                DownloadColumn.TYPE -> renderDownloadType(datenDownload)
                DownloadColumn.SOURCE -> renderDownloadSource(datenDownload)
                DownloadColumn.BUTTON_START -> handleButtonStartColumn(datenDownload, isSelected)
                DownloadColumn.BUTTON_DELETE -> handleButtonDeleteColumn(datenDownload, isSelected)
                DownloadColumn.ABO -> handleAboColumn(datenDownload)
                DownloadColumn.SENDER -> {
                    if (mvTable.showSenderIcons()) {
                        val targetDim = getSenderCellDimension(table, row, column)
                        setSenderIcon(valueText(value), targetDim, isSelected)
                    }
                }

                DownloadColumn.GEO -> datenDownload.film?.let { film -> drawGeolocationIcons(film, isSelected) }
                else -> Unit
            }

            if (downloadColumn == DownloadColumn.TITLE) {
                datenDownload.film?.let { film ->
                    setIndicatorIcons(table, film, isSelected)
                }
            }

            setBackgroundColor(this, datenDownload.runtime.runState, isSelected)
        } catch (ex: Exception) {
            logger.error(ex)
        }
        return this
    }

    private fun renderProgressColumn(datenDownload: DatenDownload, mvTable: MVTable, isSelected: Boolean): Component? {
        progressBar.border = if (mvTable.showSenderIcons() && !mvTable.getUseSmallSenderIcons()) {
            largeBorder
        } else {
            emptyBorder
        }

        val start = datenDownload.runtime.runState
        if (start == null) {
            text = ""
            return null
        }

        if (1 < start.percent && start.percent < DownloadRunState.PROGRESS_FERTIG) {
            setBackgroundColor(panel, start, isSelected)
            setBackgroundColor(progressBar, start, isSelected)

            progressBar.value = start.percent
            val progressValue = start.percent / 10.0
            progressBar.string = "$progressValue%"

            return panel
        }

        text = DownloadProgressText.getTextProgress(datenDownload.isDownloadManager, start)
        return null
    }

    private fun hideZeroFilmNumber(table: JTable, rowModelIndex: Int) {
        if (table.model.getValueAt(rowModelIndex, DownloadColumn.FILM_NUMBER.index) as Int == 0) {
            text = ""
        }
    }

    private fun renderDownloadType(datenDownload: DatenDownload) {
        text = datenDownload.art.label
    }

    private fun renderDownloadSource(datenDownload: DatenDownload) {
        text = datenDownload.quelle.label
    }

    private fun createTextArea(
        value: String,
        datenDownload: DatenDownload,
        column: DownloadColumn,
        isSelected: Boolean,
    ): JTextArea {
        val textArea = createWrappedTextArea(value)
        if (column == DownloadColumn.ABO) {
            handleAboColumn(textArea, datenDownload)
        }
        setBackgroundColor(textArea, datenDownload.runtime.runState, isSelected)
        return textArea
    }

    private fun setIconsAndToolTips(datenDownload: DatenDownload, isSelected: Boolean) {
        val start = datenDownload.runtime.runState
        if (start != null && !datenDownload.isDownloadManager) {
            when (start.status) {
                StartStatus.FINISHED -> {
                    icon = filmStartIcons.icon(isSelected)
                    toolTipText = PLAY_DOWNLOADED_FILM
                }

                StartStatus.ERROR -> {
                    icon = downloadStartIcons.icon(isSelected)
                    toolTipText = DOWNLOAD_STARTEN
                }

                else -> {
                    icon = downloadStopIcons.icon(isSelected)
                    toolTipText = DOWNLOAD_STOPPEN
                }
            }
        } else {
            icon = downloadStartIcons.icon(isSelected)
            toolTipText = DOWNLOAD_STARTEN
        }
    }

    private fun handleButtonStartColumn(datenDownload: DatenDownload, isSelected: Boolean) {
        horizontalAlignment = CENTER
        setIconsAndToolTips(datenDownload, isSelected)
    }

    private fun handleAboColumn(a: JTextArea, datenDownload: DatenDownload) {
        if (datenDownload.isFromAbo) {
            a.foreground = MVColor.DOWNLOAD_IST_ABO.color
        } else {
            a.foreground = MVColor.DOWNLOAD_IST_DIREKTER_DOWNLOAD.color
            a.text = "Download"
        }
    }

    private fun handleAboColumn(datenDownload: DatenDownload) {
        horizontalAlignment = CENTER
        if (datenDownload.isFromAbo) {
            foreground = MVColor.DOWNLOAD_IST_ABO.color
        } else {
            foreground = MVColor.DOWNLOAD_IST_DIREKTER_DOWNLOAD.color
            text = "Download"
        }
    }

    private fun handleButtonDeleteColumn(datenDownload: DatenDownload, isSelected: Boolean) {
        horizontalAlignment = CENTER
        val start = datenDownload.runtime.runState
        if (start != null) {
            if (start.status >= StartStatus.FINISHED) {
                setIcon(downloadClearIcons, DOWNLOAD_ENTFERNEN, isSelected)
            } else {
                setupDownloadLoeschen(isSelected)
            }
        } else {
            setupDownloadLoeschen(isSelected)
        }
    }

    private fun setIcon(icons: RendererIconPair, tooltip: String, isSelected: Boolean) {
        icon = icons.icon(isSelected)
        toolTipText = tooltip
    }

    private fun setupDownloadLoeschen(isSelected: Boolean) {
        setIcon(downloadDeleteIcons, DOWNLOAD_LOESCHEN, isSelected)
    }

    private companion object {
        private const val DOWNLOAD_STARTEN = "Download starten"
        private const val DOWNLOAD_LOESCHEN = "Download aus Liste entfernen"
        private const val DOWNLOAD_STOPPEN = "Download stoppen"
        private const val DOWNLOAD_ENTFERNEN = "Download entfernen"
        private const val PLAY_DOWNLOADED_FILM = "gespeicherten Film abspielen"
        private val logger = LogManager.getLogger(CellRendererDownloads::class.java)
    }
}
